sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+        eigenvalue used for the Jordan block
+        defaults to values -1 or 1
+
+    See Also
+    ========
+    jordan_normal
+
+    """
+
+    index = index or _sample(range(dim), 2)
+    scalar = _sample(_elementary_scalars) \
+        if scalar is None else scalar
+    start, end = sorted(index)
+    obj = eye(dim)
+    obj[start, start] = scalar
+    for i in range(start + 1, end):
+        obj[i - 1, i] = 1
+        obj[i, i] = scalar
+    return obj
+
+
+def transposition(dim,
+                  index=None):
+    r"""n x n transposition matrix
+
+    Explanation
+    ===========
+
+    A transposition matrix is an identity matrix where two columns (or rows)
+    are swapped. It is a special permutation matrix.
+    Moreover, any permutaion matrix is a product of transposition matrices.
+
+    In two dimensions it is
+
+    .. math::
+
+       \mathbf{T} = \left[\begin{array}{cc}0 & 1 \\ 1 & 0\end{array}\right] \\
+
+    If index is given, it sets which column or row will be swapped.
+
+    Examples
+    ========
+
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import transposition
+    >>> seed(1)
+
+    >>> transposition(4)
+    Matrix([
+    [1, 0, 0, 0],
+    [0, 0, 1, 0],
+    [0, 1, 0, 0],
+    [0, 0, 0, 1]])
+
+    >>> transposition(3, (0,1))  # no more random
+    Matrix([
+    [0, 1, 0],
+    [1, 0, 0],
+    [0, 0, 1]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+    index : tuple(integer, integer) (optional)
+        coordinates (i,j) of transposition
+    See Also
+    ========
+    elementary
+    permutation
+
+    """
+
+    index = index or _sample(range(dim), 2)
+    return super_elementary_matrix(dim, index)
+
+
+def permutation(dim,
+                perm=None):
+    r"""permutation matrix n x n
+
+    Explamation
+    ===========
+
+    A permutation matrix is a matrix with only 0 or 1 entries
+    and each column and each row consists of only one non-zero entry.
+
+    Such a matrix can be obtained by shuffling the rows
+    of an identiy matrix .i.e ``permutation(dim, perm)`` is eqivalent to
+    ``eye(dim).permute(perm)``.
+
+    Examples
+    ========
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import permutation
+    >>> seed(1)
+
+    >>> permutation(3)
+    Matrix([
+    [1, 0, 0],
+    [0, 0, 1],
+    [0, 1, 0]])
+
+    >>> permutation(3, (2,0,1))  # no more random
+    Matrix([
+    [0, 0, 1],
+    [1, 0, 0],
+    [0, 1, 0]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+    perm : tuple or list of integer (optional)
+        permutation as list of n integers, e.g. ``[2,1,0,3]`` as a
+        permutation of ``[0,1,2,3]``
+
+    See Also
+    ========
+    transposition
+
+    """
+
+    perm = perm or _sample(range(dim), dim)
+    obj = zeros(dim)
+    for i, j in enumerate(perm):
+        obj[i, j] = 1
+    return obj  # aka eye(dim).permute(perm)
```

Comment:
Is there a reason this shouldn't just use `eye(dim).permute(perm)`?

Reply:
Ok, fine. Changed.